Use register based circuit

robinhundt opened the pull request:
This commit adapts Polytune to use the `register_circuit::Circuit` of garble. This allows a more memory efficient protocol execution, as intermediate values of the computation are overwritten once they're not needed anymore.

This is part of the work on reducing the peak memory #113.

kisakishy commented on src/mpc/protocol.rs:
```diff
@@ -137,32 +138,32 @@ impl From<faand::Error> for Error {
-    InvalidInputMacOnWire(usize),
-    /// Two different parties tried to provide an input mask for the wire.
+    /// No secret share was sent during preprocessing for the specified instruction.
+    MissingPreprocessingShareForInst(Inst),
```
Hmm, I am not 100% sure I agree I agree with the renaming here. What is meant by instruction here? AND/XOR, i.e., the gates? And is this really no share for the instruction or for the result of the instruction, i.e., the wire? I am a bit confused by the wording instruction.

robinhundt replied:
I tried to remove the references to gates and wires in this new implementation, as these concepts don't exist in the new circuit representation anymore. There is a mapping between the previous gates and instructions, and wires and registers, but didn't want to use too much related but not quite same terminology.

I didn't remove the notion of gates and wires from the user facing documentation, as this is what users are probably more familiar with, at least if they have some knowledge of MPC. However, because the new circuit type is also part of the public API we should probably mention instructions and wires there as well and how it maps to gates and wires.

In the case of this specific error variant, I'd argue that this is the correct name. It is emitted here:
https://github.com/sine-fdn/polytune/blob/b2017d280c760fb49b511e58e0487cf04f4d8555/src/mpc/protocol.rs#L410-L413
The `Inst` part could be split into an `Input` and `And` variant, but I don't think this is necessary, as the relevant instruction is part of the error. We could also include the position of this instruction in the `Inst` vector, as otherwise it is not 100% clear which one is the problem. This random share is used to initialize the `shares[inst.out]` position, where `inst.out` is a register. However, the important thing is that a random share is missing for this instruction, and not this register.

Does this clear things up?
I'll also add a commit with some more explanation in the docs and add the instruction position to the error.

kisakishy reviewed (commented):
Generally it looks good to me, I was only buffled by the renaming of gates and wires to instructions. @robinhundt Could you elaborate on that a bit?

robinhundt commented:
Ok, I added a section in the docs on the gates/instructions terminology.
I chose to remove the `Inst` from the error variant and only add its index. Most of the other error variants also only have the index (in some cases like the inputs, this is the only thing we currently have in some places in the code as we're not iterating over the instructions at that point). This way it is consistent.

I notice that there are some `unwrap`s in some places which should be unreachable. I've changed these to propagate errors in #204 already.
